bindings/rust/src/lib.rs

```diff
@@ -755,4 +824,246 @@ mod tests {
+        unsafe {
+            assert!(instance.checked_memory_slice(0, 0).is_err());
+            assert!(instance.checked_memory_slice_mut(0, 0).is_err());
+            assert!(instance.checked_memory_slice(0, 65536).is_err());
+            assert!(instance.checked_memory_slice_mut(0, 65536).is_err());
+            assert!(instance.checked_memory_slice(65535, 1).is_err());
+            assert!(instance.checked_memory_slice_mut(65535, 1).is_err());
+            assert!(instance.checked_memory_slice(65535, 2).is_err());
+            assert!(instance.checked_memory_slice_mut(65535, 2).is_err());
+            assert!(instance.checked_memory_slice(65536, 0).is_err());
+            assert!(instance.checked_memory_slice_mut(65536, 0).is_err());
+        }
+
+        // Set memory via safe helper.
+        assert!(instance.memory_set(0, &[]).is_err());
+        // Get memory via safe helper.
+        let mut dst: Vec<u8> = Vec::new();
+        dst.resize(65536, 0);
+        // Reading 65536 bytes.
+        assert!(instance.memory_get(0, &mut dst).is_err());
+    }
+
+    #[test]
+    fn memory() {
+        /* wat2wasm
+        (module
+          (func (export "grow") (param i32) (result i32) (memory.grow (local.get 0)))
+          (func (export "peek") (param i32) (result i32) (i32.load (local.get 0)))
+          (func (export "poke") (param i32) (param i32) (i32.store (local.get 0) (local.get 1)))
+          (memory (export "mem") 1 2)
+        )
+        */
+        let input = hex::decode("0061736d01000000010b0260017f017f60027f7f00030403000001050401010102071c040467726f770000047065656b000104706f6b650002036d656d02000a1a030600200040000b070020002802000b0900200020013602000b").unwrap();
+        let module = parse(&input);
+        assert!(module.is_ok());
+        let instance = module.unwrap().instantiate();
+        assert!(instance.is_ok());
+        let mut instance = instance.unwrap();
+
+        assert_eq!(instance.memory_size(), 65536);
+        unsafe {
+            // Allow empty slices.
+            assert!(instance.checked_memory_slice(0, 0).is_ok());
+            assert!(instance.checked_memory_slice_mut(0, 0).is_ok());
+            // Entire memory.
+            assert!(instance.checked_memory_slice(0, 65536).is_ok());
+            assert!(instance.checked_memory_slice_mut(0, 65536).is_ok());
+            // Allow empty slices.
+            assert!(instance.checked_memory_slice(65535, 0).is_ok());
+            assert!(instance.checked_memory_slice_mut(65535, 0).is_ok());
+            assert!(instance.checked_memory_slice(65536, 0).is_ok());
+            assert!(instance.checked_memory_slice_mut(65536, 0).is_ok());
+            // Single byte.
+            assert!(instance.checked_memory_slice(65535, 1).is_ok());
+            assert!(instance.checked_memory_slice_mut(65535, 1).is_ok());
+            // Reading over.
+            assert!(instance.checked_memory_slice(65535, 2).is_err());
```

Comment:
`assert!(instance.checked_memory_slice(65536, 1).is_err());` could be added

Reply:
Done.